### PR TITLE
Categories crud

### PR DESCRIPTION
**Pull request by UtsavSomaiya**
- [ ] In this PR we can perform category CRUD.

### CODE REVIEW

File: app/Models/Category.php [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+class Category extends Authenticatable
+{
+    use HasFactory;
+
+    protected $fillable = ['name'];
+
+    public function products()
+    {
+        return $this->hasMany(Product::class);
+    }
```
Comment (hirenkeraliya) [inline app/Models/Category.php:17]: remove this

File: app/Http/Controllers/Admin/CategoryController.php [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        ]);
+
+        Category::create($category);
+
+        return to_route('categories-list')->with(
+            ['success' => 'Category added successfully.']
+        );
+    }
+
+    public function show()
```
Comment (hirenkeraliya) [inline app/Http/Controllers/Admin/CategoryController.php:24]: index it's not an show

Comment (hirenkeraliya) [reply app/Http/Controllers/Admin/CategoryController.php:24]: same rename everywhere

File: resources/views/admin/categories/show.blade.php [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+                        <thead>
+                            <tr>
+                                <th>Id</th>
+                                <th>Name</th>
+                                <th colspan="2">Action</th>
+                            </tr>
+                        </thead>
+                        <tbody>
+                            @foreach($categories as $category)
+                                <tr>
```
Comment (hirenkeraliya) [inline resources/views/admin/categories/show.blade.php:25]: use forelse

**Review by hirenkeraliya — CHANGES_REQUESTED** [review]

File: app/Http/Controllers/Admin/CategoryController.php [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+{
+    public function store(Request $request)
+    {
+        $category = $request->validate([
+            'name' => ['required','min:3','max:255','unique:categories,name']
+        ]);
+
+        Category::create($category);
+
+        return to_route('categories-list')->with(
```
[the remark this answers is not in the document]
Comment (UtsavSomaiya) [reply app/Http/Controllers/Admin/CategoryController.php:19]: laravel url(route) name => `rajan-parmar` (General Case)

File: app/Http/Controllers/Admin/CategoryController.php [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        ]);
+
+        Category::create($category);
+
+        return to_route('categories_list')->with(
+            ['success' => 'Category added successfully.']
+        );
+    }
+
+    public function index()
```
Comment (hirenkeraliya) [inline app/Http/Controllers/Admin/CategoryController.php:24]: index method pela aave pa6i store edit update delete

File: app/Http/Controllers/Admin/CategoryController.php [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    {
+        $category = Category::findOrFail($categoryId)->first();
+
+        return view('admin.categories.add', compact('category'));
+    }
+
+    public function update($categoryId, Request $request)
+    {
+        $category = $request->validate([
+            'name' => ['required','min:3','max:255']
```
Comment (hirenkeraliya) [inline app/Http/Controllers/Admin/CategoryController.php:50]: Unique validation?

File: app/Http/Controllers/Admin/CategoryController.php [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+        Category::findOrFail($categoryId)->delete();
+
+        return back()->with([
+            'success' => 'Category deleted successfully.'
+        ]);
+    }
+
+    public function edit($categoryId)
+    {
+        $category = Category::findOrFail($categoryId)->first();
```
Comment (hirenkeraliya) [inline app/Http/Controllers/Admin/CategoryController.php:42]: `->first()` not needed.

File: database/migrations/2022_05_26_070834_create_categories_table.php [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+    /**
+     * Run the migrations.
+     *
+     * @return void
+     */
+    public function up()
+    {
+        Schema::create('categories', function (Blueprint $table) {
+            $table->id();
+            $table->string('name');
```
Comment (hirenkeraliya) [inline database/migrations/2022_05_26_070834_create_categories_table.php:17]: need to declare unique

File: resources/views/admin/categories/add.blade.php [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+                        @enderror
+                    </div>
+                    <button type="submit" class="btn btn-primary me-2 mt-3" name="submit">
+                        @isset($category)
+                            Update
+                        @else
+                            Submit
+                        @endisset
+                    </button>
+                    <a href="{{ route('categories_list') }}" class="btn btn-light mt-3">Cancel</a>
```
Comment (hirenkeraliya) [inline resources/views/admin/categories/add.blade.php:33]: only categories

File: resources/views/admin/categories/index.blade.php [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+                            <th colspan="2">Action</th>
+                        </tr>
+                    </thead>
+                    <tbody>
+                        @forelse($categories as $category)
+                            <tr>
+                                <td>{{ $category->id }}</td>
+                                <td>{{ $category->name }}</td>
+                                <td>
+                                    <a href="{{ route('category_edit', ['id' => $category->id ]) }}"
```
Comment (hirenkeraliya) [inline resources/views/admin/categories/index.blade.php:28]: categories.edit

File: resources/views/admin/categories/index.blade.php [hunk 11]
```diff
@@ -0,0 +1,59 @@
+@extends('admin.layout')
+@section('content')
+    <div class="col-md-8">
+        <div class="card border-0 shadow-lg">
+            <div class="card-title">
+                <div class="mt-2">
+                    <span class="h4 ms-3 me-5">Categories</span>
+                    <span class="ms-3">
+                        <a href="{{ route('category_add') }}" class="link-dark h6">Add New Category</a>
```
Comment (hirenkeraliya) [inline resources/views/admin/categories/index.blade.php:9]: categories.add

File: resources/views/admin/categories/index.blade.php [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+                                <td>
+                                    <a href="{{ route('category_edit', ['id' => $category->id ]) }}"
+                                        class="link-dark"
+                                    >
+                                        <i class="fa-solid fa-pencil"></i>
+                                    </a>
+                                </td>
+                                <td>
+                                    <form method="post"
+                                        action="{{ route('category_delete', ['id' => $category->id]) }}"
```
Comment (hirenkeraliya) [inline resources/views/admin/categories/index.blade.php:36]: categories.delete

File: resources/views/admin/categories/index.blade.php [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+                    </span>
+                </div>
+            </div>
+            <div class="card-body">
+                <table class="table w-50">
+                    <thead>
+                        <tr>
+                            <th>Id</th>
+                            <th>Name</th>
+                            <th colspan="2">Action</th>
```
Comment (hirenkeraliya) [inline resources/views/admin/categories/index.blade.php:19]: why colspan in the header?

**Review by hirenkeraliya — CHANGES_REQUESTED** [review]

File: resources/views/admin/categories/add.blade.php [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+                        @error('name')
+                            <label class="text-danger">{{ $message }}</label>
+                        @enderror
+                    </div>
+                    <button type="submit" class="btn btn-primary me-2 mt-3" name="submit">
+                        @isset($category)
+                            Update
+                        @else
+                            Submit
+                        @endisset
```
Comment (hirenkeraliya) [inline resources/views/admin/categories/add.blade.php:31]: `{{ $category ? 'Update' : 'Submit' }}`

File: routes/web.php [hunk 15]
```diff
@@ -24,6 +25,13 @@
...
 Route::middleware('auth')->group(function () {
     Route::view('/admin/dashboard', 'admin.dashboard')->name('dashboard');
-    
+
+    Route::view('/admin/categories/add', 'admin.categories.add')->name('categories.add');
+    Route::post('admin/categories/add', [CategoryController::class,'store']);
+    Route::get('/admin/categories', [CategoryController::class,'index'])->name('categories');
+    Route::post("admin/categories/delete/{id}", [CategoryController::class,'delete'])->name('categories.delete');
+    Route::get("admin/categories/edit/{id}", [CategoryController::class,'edit'])->name('categories.edit');
+    Route::post("admin/categories/edit/{id}", [CategoryController::class,'update']);
```
Comment (hirenkeraliya) [inline routes/web.php:34]:
```suggestion
    Route::get('/admin/categories', [CategoryController::class,'index'])->name('categories');
    Route::view('/admin/categories/add', 'admin.categories.add')->name('categories.add');
    Route::post('admin/categories', [CategoryController::class,'store']);
    Route::get("admin/categories/edit/{category}", [CategoryController::class,'edit'])->name('categories.edit');
    Route::put("admin/categories/edit/{category}", [CategoryController::class,'update']);
    Route::post("admin/categories/delete/{id}", [CategoryController::class,'delete'])->name('categories.delete');
```

File: app/Http/Controllers/Admin/CategoryController.php [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+        Category::findOrFail($categoryId)->update($category);
+
+        return to_route('categories')->with([
+            'success' => 'Category Updated successfully'
+        ]);
+    }
+
+    public function delete($categoryId)
+    {
+        Category::findOrFail($categoryId)->delete();
```
Comment (hirenkeraliya) [inline app/Http/Controllers/Admin/CategoryController.php:59]: you can use model binding like

```suggestion
    public function delete(Category $category)
    {
        $category->delete();
```

**Review by hirenkeraliya — CHANGES_REQUESTED** [review]

File: resources/views/admin/categories/add.blade.php (hunk 14 above)
Comment (UtsavSomaiya) [reply resources/views/admin/categories/add.blade.php:31]: `{{ isset($category) ? 'Update' : 'Submit' }}`

File: routes/web.php [hunk 17]
```diff
@@ -24,6 +25,14 @@
 
 Route::middleware('auth')->group(function () {
     Route::view('/admin/dashboard', 'admin.dashboard')->name('dashboard');
-    
+
+    Route::get('/admin/categories', [CategoryController::class,'index'])->name('categories');
+    Route::view('/admin/categories/add', 'admin.categories.add')->name('categories.add');
+    Route::post('admin/categories', [CategoryController::class,'store']);
```
Comment (hirenkeraliya) [inline routes/web.php:31]: name should be `categories.store` and use it on the form

**Review by hirenkeraliya — CHANGES_REQUESTED** [review]

**Review by hirenkeraliya — APPROVED** [review]

**Pull request merged**